aws-opentelemetry-distro/src/amazon/opentelemetry/distro/exporter/aws/metrics/_cloudwatch_log_client.py

```diff
@@ -0,0 +1,354 @@
+
+            elapsed_ms = int((time.time() - start_time) * 1000)
+            logger.debug(
+                "Successfully sent %s log events (%s KB) in %s ms",
+                len(batch["logEvents"]),
+                batch["byteTotal"] / 1024,
+                elapsed_ms,
+            )
+
+            return response
+
+        except ClientError as error:
+            # Handle resource not found errors by creating log group/stream
+            error_code = error.response.get("Error", {}).get("Code")
+            if error_code == "ResourceNotFoundException":
+                logger.info("Log group or stream not found, creating resources and retrying")
+
+                try:
+                    # Create log group first
+                    self._create_log_group_if_needed()
+                    # Then create log stream
+                    self._create_log_stream_if_needed()
+
+                    # Retry the PutLogEvents call
+                    response = self.logs_client.put_log_events(**put_log_events_input)
+
+                    elapsed_ms = int((time.time() - start_time) * 1000)
+                    logger.debug(
+                        "Successfully sent %s log events (%s KB) in %s ms after creating resources",
+                        len(batch["logEvents"]),
+                        batch["byteTotal"] / 1024,
+                        elapsed_ms,
+                    )
+
+                    return response
+
+                except ClientError as retry_error:
+                    logger.error("Failed to send log events after creating resources: %s", retry_error)
+                    raise
+            else:
+                logger.error("Failed to send log events: %s", error)
+                raise
+
+    def send_log_event(self, log_event: Dict[str, Any]):
+        """
+        Send a log event to CloudWatch Logs.
+
+        This function implements the same logic as the Go version in the OTel Collector.
+        It batches log events according to CloudWatch Logs constraints and sends them
+        when the batch is full or spans more than 24 hours.
+
+        Args:
+            log_event: The log event to send
+        """
+        try:
+            # Validate the log event
+            if not self._validate_log_event(log_event):
+                return
+
+            # Calculate event size
+            event_size = len(log_event["message"]) + self.CW_PER_EVENT_HEADER_BYTES
+
+            # Initialize event batch if needed
+            if self._event_batch is None:
+                self._event_batch = self._create_event_batch()
+
+            # Check if we need to send the current batch and create a new one
+            current_batch = self._event_batch
+            if self._event_batch_exceeds_limit(current_batch, event_size) or not self._is_batch_active(
+                current_batch, log_event["timestamp"]
+            ):
+                # Send the current batch
+                self._send_log_batch(current_batch)
```

Comment:
What happens if the current batch export fails? The current logic assumes it succeeds and will proceed to create a new batch. Is that the expected behavior?

Reply:
NVM. I see that the `_send_log_batch` will raise exception if export fails.